federationapi/routing/join.go

```diff
@@ -33,153 +34,144 @@ import (
+			return nil, nil, &util.JSONResponse{
+				Code: http.StatusNotFound,
+				JSON: spec.NotFound("Room does not exist"),
+			}
+		} else if e, ok := err.(gomatrixserverlib.BadJSONError); ok {
+			return nil, nil, &util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: spec.BadJSON(e.Error()),
+			}
+		} else if err != nil {
+			util.GetLogger(httpReq.Context()).WithError(err).Error("eventutil.BuildEvent failed")
+			e := spec.InternalServerError()
+			return nil, nil, &e
 		}
-	}
 
-	// Check if we think we are still joined to the room
-	inRoomReq := &api.QueryServerJoinedToRoomRequest{
-		ServerName: cfg.Matrix.ServerName,
-		RoomID:     roomID,
-	}
-	inRoomRes := &api.QueryServerJoinedToRoomResponse{}
-	if err = rsAPI.QueryServerJoinedToRoom(httpReq.Context(), inRoomReq, inRoomRes); err != nil {
-		util.GetLogger(httpReq.Context()).WithError(err).Error("rsAPI.QueryServerJoinedToRoom failed")
-		return spec.InternalServerError()
-	}
-	if !inRoomRes.RoomExists {
-		return util.JSONResponse{
-			Code: http.StatusNotFound,
-			JSON: spec.NotFound(fmt.Sprintf("Room ID %q was not found on this server", roomID)),
-		}
-	}
-	if !inRoomRes.IsInRoom {
-		return util.JSONResponse{
-			Code: http.StatusNotFound,
-			JSON: spec.NotFound(fmt.Sprintf("Room ID %q has no remaining users on this server", roomID)),
+		stateEvents := make([]gomatrixserverlib.PDU, len(queryRes.StateEvents))
+		for i, stateEvent := range queryRes.StateEvents {
+			stateEvents[i] = stateEvent.PDU
 		}
+		return event, stateEvents, nil
 	}
 
-	// Check if the restricted join is allowed. If the room doesn't
-	// support restricted joins then this is effectively a no-op.
-	res, authorisedVia, err := checkRestrictedJoin(httpReq, rsAPI, roomVersion, roomID, userID)
-	if err != nil {
-		util.GetLogger(httpReq.Context()).WithError(err).Error("checkRestrictedJoin failed")
-		return spec.InternalServerError()
-	} else if res != nil {
-		return *res
+	roomQuerier := JoinRoomQuerier{
+		roomserver: rsAPI,
 	}
 
-	// Try building an event for the server
-	proto := gomatrixserverlib.ProtoEvent{
-		Sender:   userID,
-		RoomID:   roomID,
-		Type:     "m.room.member",
-		StateKey: &userID,
+	input := gomatrixserverlib.HandleMakeJoinInput{
+		Context:            httpReq.Context(),
+		UserID:             userID,
+		RoomID:             roomID,
+		RoomVersion:        roomVersion,
+		RemoteVersions:     remoteVersions,
+		RequestOrigin:      request.Origin(),
+		RequestDestination: request.Destination(),
+		LocalServerName:    cfg.Matrix.ServerName,
+		RoomQuerier:        &roomQuerier,
+		BuildEventTemplate: createJoinTemplate,
 	}
-	content := gomatrixserverlib.MemberContent{
-		Membership:    spec.Join,
-		AuthorisedVia: authorisedVia,
-	}
-	if err = proto.SetContent(content); err != nil {
-		util.GetLogger(httpReq.Context()).WithError(err).Error("builder.SetContent failed")
-		return spec.InternalServerError()
+	response, internalErr := gomatrixserverlib.HandleMakeJoin(input)
```

Comment:
I'd rather this function wasn't so implementation specific (it heavily implies the /make_join endpoint, but also presumably does /send_join despite it not saying so?) Can we just call it `JoinRoom(input)`?

Reply:
This is on the receiving side and only deals with handling the /make_join endpoint.
The /send_join handling is completely separate.